src/v/cluster/rm_stm.cc

```diff
@@ -1196,4 +1277,53 @@ ss::future<stm_snapshot> rm_stm::take_snapshot() {
     co_return stx_ss;
 }
 
+ss::future<> rm_stm::save_abort_snapshot(abort_snapshot snapshot) {
+    iobuf snapshot_data;
+    reflection::adl<abort_snapshot>{}.to(snapshot_data, snapshot);
+    uint64_t snapshot_size = snapshot_data.size_bytes();
+
+    auto filename = fmt::format(
+      "abort.idx.{}.{}", snapshot.first, snapshot.last);
+    auto writer = co_await _abort_snapshot_mgr.start_snapshot(filename);
+
+    iobuf metadata_buf;
+    reflection::serialize(metadata_buf, abort_snapshot_version, snapshot_size);
+    co_await writer.write_metadata(std::move(metadata_buf));
+    co_await write_iobuf_to_output_stream(
+      std::move(snapshot_data), writer.output());
+    co_await writer.close();
+    co_await _abort_snapshot_mgr.finish_snapshot(writer);
+}
+
+ss::future<std::optional<rm_stm::abort_snapshot>>
+rm_stm::load_abort_snapshot(abort_index index) {
+    auto filename = fmt::format("abort.idx.{}.{}", index.first, index.last);
+
+    auto reader = co_await _abort_snapshot_mgr.open_snapshot(filename);
+    if (!reader) {
+        co_return std::nullopt;
+    }
+
+    auto meta_buf = co_await reader->read_metadata();
+    iobuf_parser meta_parser(std::move(meta_buf));
+
+    auto version = reflection::adl<int8_t>{}.from(meta_parser);
+    vassert(
+      version == abort_snapshot_version,
+      "Only support abort_snapshot_version {} but got {}",
+      abort_snapshot_version,
+      version);
+
+    auto snapshot_size = reflection::adl<int32_t>{}.from(meta_parser);
```

Comment:
did you serialize the size before as a 64-bit integer?

Reply:
it's often a good idea to add an assertion that the iobuf was fully consume after parsing (e.g. `vassert(reader.bytes_left() == 0)` you'll have to check the exact api) which would catch this issue.

Reply:
right, it's a bag, changed in both places to be int32_t